torch_np/_wrapper.py

```diff
@@ -311,16 +311,7 @@ def full(shape, fill_value, dtype=None, order="C", *, like=None):
     if isinstance(fill_value, ndarray):
         fill_value = fill_value.get()
     if dtype is None:
-        if isinstance(fill_value, bool):
-            torch_dtype = torch.bool
-        elif isinstance(fill_value, int):
-            torch_dtype = torch.int64
-        elif isinstance(fill_value, float):
-            torch_dtype = torch.float64
-        elif isinstance(fill_value, complex):
-            torch_dtype = torch.complex128
-        else:
-            torch_dtype = _dtypes.torch_dtype_from(dtype)
+        torch_dtype = _dtypes.torch_dtype_from(type(fill_value))
```

Comment:
Sadly, this won't work because 'scalar' here can actually a zero-dim torch tensor (e.g. in reductions -- cf CI failures) and 

```
(Pdb) p type(arg)
<class 'torch._C._TensorMeta'>
(Pdb) torch.is_tensor(arg)
```

I'd suggest to roll back the last commit: I'm going to work through dtype inference anyway, so as long as the behavior is correct, I'll tidy it up. Unless you want to take a detour and tidy up the dtype handling yourself of course :-).

The usage of `np.full` in reductions is arguably a bug itself, so maybe it should be fixed in there. At any rate, this change is great that it shows that the current infra is too brittle and shows its breaking points. Which is exactly what's needed, getting data on actual usage.

Reply:
Ah I forgot you can pass 0d arrays as the `fill_value`, nice catch—I implemented your fix and updated the introduced `test_full` test to cover this.